04-vue-cli/02-TheToaster/components/TheToaster.vue

```diff
@@ -1,25 +1,40 @@
 <template>
   <div class="toasts">
-    <div class="toast toast_success">
-      <UiIcon class="toast__icon" icon="check-circle" />
-      <span>Success Toast Example</span>
-    </div>
-
-    <div class="toast toast_error">
-      <UiIcon class="toast__icon" icon="alert-circle" />
-      <span>Error Toast Example</span>
-    </div>
+    <TheToast v-for="toast in toasts" :kind="toast.kind" :text="toast.text" :key="toast.text" @close="update" />
   </div>
 </template>
 
-<script>
-import UiIcon from './UiIcon.vue';
+<script lang="ts">
+import { defineComponent } from 'vue';
+import TheToast, { EToastKind } from './TheToast.vue';
+
+type TToasterData = { toasts: { kind: EToastKind; text: string }[] };
 
-export default {
+export default defineComponent({
   name: 'TheToaster',
 
-  components: { UiIcon },
-};
+  components: { TheToast },
+
+  data(): TToasterData {
+    return {
+      toasts: [],
+    };
+  },
+
+  methods: {
+    success(text: string) {
+      this.toasts.push({ kind: EToastKind.SUCCESS, text });
+    },
+
+    error(text: string) {
+      this.toasts.push({ kind: EToastKind.ERROR, text });
+    },
+
+    update(text: string) {
+      this.toasts = this.toasts.filter((it) => it.text !== text);
+    },
```

Comment:
The text cannot be toast's identifier. It is not guaranteed to be unequal.
For example, often the toast is just "Saved".

Try to make 10 toasts at a different time with the text "Saved". All will be closed at the same time.

Reply:
By the way, in Vue we don't have to change a component's state immutably. 

Removing one element with the `splice` method is faster than recreating an array after filtering.

Moreover, we don't need to search for the element here. This method is bound as an event handler in `v-for`. We know exactly what element by the index was removed.